Adding hydra remote start (#25)
Constellation-Labs/euclid-development-environment: README.md

## Deployment

> Configuring, deploying, and starting to remote node instances is supported through Ansible playbooks. The default settings deploy to three node instances via SSH which host all layers of your metagraph project (gL0, mL0, cL1, dL1). Two hydra methods are available to help with the deployment process: `hydra remote_deploy` and `hydra remote_start`.

Reviewer: ```suggestion
Configuring, deploying, and starting remote node instances is supported through Ansible playbooks. The default settings deploy to three node instances via SSH which host all layers of your metagraph project (gL0, mL0, cL1, dL1). Two hydra methods are available to help with the deployment process: `hydra remote_deploy` and `hydra remote_start`.
```

Author: Fixed

Prior to running these methods, remote host information must be configured in  `infra/ansible/hosts.ansible.yml`

### `hydra remote_deploy`

> This method configures remote instances with all the necessary dependencies to run a Metagraph, including Java, Scala, and required build tools. The Ansible playbook used for this process can be found and edited in `infra/ansible/playbooks/deploy.ansible.yml`. Also, creates all required directories on the remote hosts, and creates or updates metagraph files to match your local Euclid environment. Specifically, it creates the following directories:

Reviewer: "metagraph" shouldn't be capitalized

```suggestion
This method configures remote instances with all the necessary dependencies to run a metagraph, including Java, Scala, and required build tools. The Ansible playbook used for this process can be found and edited in `infra/ansible/playbooks/deploy.ansible.yml`. It also creates all required directories on the remote hosts, and creates or updates metagraph files to match your local Euclid environment. Specifically, it creates the following directories:
```

Author: Fixed

## Deployment

Configuring, deploying, and starting remote node instances is supported through Ansible playbooks. The default settings deploy to three node instances via SSH which host all layers of your metagraph project (gL0, mL0, cL1, dL1). Two hydra methods are available to help with the deployment process: `hydra remote_deploy` and `hydra remote_start`.
Prior to running these methods, remote host information must be configured in  `infra/ansible/hosts.ansible.yml`

### Host Configuration

> To run your metagraph remotely, you'll need remote hosts—ideally, at least three of them. These hosts should be running either `ubuntu-20.04` or `ubuntu-22.04`. It's recommended that each host meets the following minimum requirements:

Reviewer: ```suggestion
To run your metagraph remotely, you'll need remote server instances - 3 instances for the default configuration. These hosts should be running either `ubuntu-20.04` or `ubuntu-22.04`. It's recommended that each host meets the following minimum requirements:
```

-   16GB of RAM
-   8vCPU
-   160GB of storage

You can choose your preferred platform for hosting your instances, such as AWS or DigitalOcean. After creating your hosts, you'll need to provide the following information in the `hosts.ansible.yml` file:

-   Host IP
-   Host user

> -   Host SSH key (optional if your localhost already has access to the remote host)

Reviewer: ```suggestion
-   Host SSH key (optional if your default SSH token already has access to the remote host)
```

### P12 Files

> These files serve as your token files and should be located in the `source/p12-files` directory by default. Details such as the `file-name`, `key-alias`, and `password` should be specified in the `euclid.json` file under the `p12_files` section. By default, Euclid comes with three example files: `token-key.p12`, `token-key-1.p12`, and `token-key-2.p12`. **NOTE:** Before deploying, be sure to replace these example files with your own, as these files are public and their credentials are shared.

Reviewer: ```suggestion
P12 files contain the public/private key pair identifying each node (peerID) and should be located in the `source/p12-files` directory by default. The `file-name`, `key-alias`, and `password` should be specified in the `euclid.json` file under the `p12_files` section. By default, Euclid comes with three example files: `token-key.p12`, `token-key-1.p12`, and `token-key-2.p12`. **NOTE:** Before deploying, be sure to replace these example files with your own, as these files are public and their credentials are shared.
```

> Each `p12` file contains a `peerId`, and each node is associated with a `peerId`. **NOTE:** Ensure that your peerIDs are registered on the MainNet network. Otherwise, the metagraph startup will fail because the network will reject the snapshots.

Reviewer: ```suggestion
**NOTE:** If deploying to MainNet, ensure that your peerIDs are registered and present on the metagraph seedlist. Otherwise, the metagraph startup will fail because the network will reject the snapshots.
```

### Network Selection

> Currently, there are three networks available for running your metagraph: `TestNet`, `IntegrationNet`, and `MainNet`. You need to specify the network on which your metagraph will run in the `euclid.json` file under `deploy -> network -> name`.

Reviewer: ```suggestion
Currently, there are two networks available for running your metagraph: `IntegrationNet`, and `MainNet`. You need to specify the network on which your metagraph will run in the `euclid.json` file under `deploy -> network -> name`.
```

### GL0 Node Configuration

> The deploy script does not deploy the `gl0` node. It's recommended to use `nodectl` to build your `gl0` node. Information on installing `nodectl` can be found [here](https://docs.constellationnetwork.io/validate/automated/nodectl). `Nodectl` helps manage `gl0` nodes by providing tools such as `auto-upgrade` and `auto-restart`.

Reviewer: ```suggestion
The deploy script does not deploy the `gl0` node. It's recommended to use `nodectl` to build your `gl0` node. Information on installing `nodectl` can be found [here](https://docs.constellationnetwork.io/validate/automated/nodectl). `Nodectl` helps manage `gl0` nodes by providing tools such as `auto-upgrade` and `auto-restart` which keep the node online in the case of a disconnection or network upgrade. Using these features is highly recommended for the stability of your metagraph. 
```

**In `code/data-l1`:**
-   data-l1.jar     // The executable for the dL1 layer

### `hydra remote_start`

> This method initiates the remote startup of your metagraph in one of the available networks: testnet, integrationnet, and mainnet.

Reviewer: We should add here that the network is set in euclid.json under deploy->network.

Author: I agree

> To commence the remote startup of the metagraph, we utilize the parameters configured on Euclid. The startup process unfolds as follows:

Reviewer: Which parameters configured in Euclid do you mean here? Maybe it could be more specific. 

```suggestion
To begin the remote startup of the metagraph, we utilize the parameters configured in Euclid. The startup process unfolds as follows:
```

Author: Got it, I'll add

> 1.  Termination of any processes currently running on the metagraph ports, which by default are 7000 for ml0, 8000 for cl1, and 9000 for dl1.

Reviewer: Maybe add that these ports can be configured in the hosts.ansible file.

Author: It's already set there
``` 
  vars:
    ansible_ssh_common_args: "-o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null"
    metagraph_l0_public_port: 7000
    metagraph_l0_p2p_port: 7001
    metagraph_l0_cli_port: 7002
    currency_l1_public_port: 8000
    currency_l1_p2p_port: 8001
    currency_l1_cli_port: 8002
    data_l1_public_port: 9000
    data_l1_p2p_port: 9001
    data_l1_cli_port: 9002
```

I'll add this information on README

> 2.  Relocation of any existing logs to a folder named `older-logs`, residing within each layer directory: `metagraph-l0`, `currency-l1`, and `data-l1`.

Reviewer: Let's change "older-logs" to something else. How about "archived-logs"?

Author: I agree

3.  Initiation of the `metagraph-l0` layer, with `node-1` designated as the genesis node.

> 4.  Initial startup as `genesis`, transitioning to `rollback` for subsequent executions. To force a genesis startup, utilize the `--force_genesis` flag with the `hydra remote_start` command.  This will move the current `data` directory to a folder named `older-datas`.

Reviewer: Let's change "older-datas" to something else. How about "archived-data"?

Author: I agree

5.  Detection of missing files required for layer execution, such as `:your_file.p12` and `metagraph-l0.jar`, triggering an error and halting execution.

> 6.  Following the initiation of `metagraph-l0`, the l1 layers, namely `currency-l1` and `data-l1`, are started. These layers only commence if their respective JARs are present in the directories. Unlike `metagraph-l0`, the absence of JARs does not prompt an error but rather prevents startup.

Reviewer: What would be the case where the jars aren't present?

Author: The layers won't start, only the metagraph-l0 will start. The snapshots will be produced but without l1 layers

After the script completes execution, you can verify if your metagraph is generating snapshots by checking the block explorer of the selected network:

-   Testnet: [https://be-testnet.constellationnetwork.io/currency/:your_metagraph_id/snapshots/latest](https://be-testnet.constellationnetwork.io/currency/:your_metagraph_id/snapshots/latest)
-   Integrationnet: [https://be-integrationnet.constellationnetwork.io/currency/:your_metagraph_id/snapshots/latest](https://be-integrationnet.constellationnetwork.io/currency/:your_metagraph_id/snapshots/latest)
-   Mainnet: [https://be-mainnet.constellationnetwork.io/currency/:your_metagraph_id/snapshots/latest](https://be-mainnet.constellationnetwork.io/currency/:your_metagraph_id/snapshots/latest)

> **NOTE:** Ensure that your peerIDs are registered on the desired network; otherwise, the metagraph startup will fail because the network will reject the snapshots.

Reviewer: For MainNet only...

**In `code/data-l1`:**
-   data-l1.jar     // The executable for the dL1 layer

### `hydra remote_start`

> This method initiates the remote startup of your metagraph in one of the available networks: testnet, integrationnet, and mainnet. The network should be set in `euclid.json` under `deploy` -> `network`

Reviewer: ```suggestion
This method initiates the remote startup of your metagraph in one of the available networks: integrationnet or mainnet. The network should be set in `euclid.json` under `deploy` -> `network`
```

To begin the remote startup of the metagraph, we utilize the parameters configured in euclid.json (`network`, `gl0_node -> ip`, `gl0_node -> id`, `gl0_node -> public_port`, `ansible -> hosts`, and `ansible -> playbooks -> start`). The startup process unfolds as follows:

1.  Termination of any processes currently running on the metagraph ports, which by default are 7000 for ml0, 8000 for cl1, and 9000 for dl1 (you can change on `hosts.ansible.yml`).
2.  Relocation of any existing logs to a folder named `archived-logs`, residing within each layer directory: `metagraph-l0`, `currency-l1`, and `data-l1`.
3.  Initiation of the `metagraph-l0` layer, with `node-1` designated as the genesis node.

> 4.  Initial startup as `genesis`, transitioning to `rollback` for subsequent executions. To force a genesis startup, utilize the `--force_genesis` flag with the `hydra remote_start` command.  This will move the current `data` directory to a folder named `archived-data`.

Reviewer: ```suggestion
4.  Initial startup as `genesis`, transitioning to `rollback` for subsequent executions. To force a genesis startup, utilize the `--force_genesis` flag with the `hydra remote_start` command.  This will move the current `data` directory to a folder named `archived-data` and restart the metagraph from the first snapshot.
```

5.  Detection of missing files required for layer execution, such as `:your_file.p12` and `metagraph-l0.jar`, triggering an error and halting execution.

> 6.  Following the initiation of `metagraph-l0`, the l1 layers, namely `currency-l1` and `data-l1`, are started. These layers only commence if their respective JARs are present in the directories. Unlike `metagraph-l0`, the absence of JARs does not prompt an error but rather prevents startup.

Reviewer: ```suggestion
6.  Following the initiation of `metagraph-l0`, the l1 layers, namely `currency-l1` and `data-l1`, are started. These layers only started if present in your project. 
```

After the script completes execution, you can verify if your metagraph is generating snapshots by checking the block explorer of the selected network:

-   Testnet: [https://be-testnet.constellationnetwork.io/currency/:your_metagraph_id/snapshots/latest](https://be-testnet.constellationnetwork.io/currency/:your_metagraph_id/snapshots/latest)
-   Integrationnet: [https://be-integrationnet.constellationnetwork.io/currency/:your_metagraph_id/snapshots/latest](https://be-integrationnet.constellationnetwork.io/currency/:your_metagraph_id/snapshots/latest)
-   Mainnet: [https://be-mainnet.constellationnetwork.io/currency/:your_metagraph_id/snapshots/latest](https://be-mainnet.constellationnetwork.io/currency/:your_metagraph_id/snapshots/latest)

> 

Reviewer: Maybe add where to check logs or API endpoints for each layer also?
